### PR TITLE
Add other tab to projects filter tabs

### LINKED ISSUE #133: Add other tab to projects filter tabs
### Description
If we have a portfolio with a lot of different project labels, it will break the layout. 
We need to allow up to 3-5 most popular labels (please check visually), and the rest of the projects with different labels should be shown under "Other" tab like on the designs

### Screenshot
Screenshot 2020-11-10 at 19 55 34

### PR DESCRIPTION
**Pull request by KlaraGajaszek**
closed #133

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,6 +40,7 @@
     "gatsby-source-filesystem": "^2.3.27",
     "gatsby-transformer-remark": "^2.8.32",
     "gatsby-transformer-sharp": "^2.5.7",
+    "lodash": "^4.17.20",
```
Comment (mecies) [inline package.json]: Please do not add `lodash` to the project, we don't need a heavy library, we only need one simple function 🎉

File: src/pages/portfolio/index.tsx [hunk 2]
```diff
@@ -75,27 +76,75 @@ const useStyles = makeStyles((theme) => ({
...
+
+const compareLengthOfProjects = (
+  longerLengthGroupedProjectsByLabel: GroupedProjectsByLabelType,
+  shorterLengthGroupedProjectsByLabel: GroupedProjectsByLabelType,
+): number => {
+  return (
+    shorterLengthGroupedProjectsByLabel.projectProperties.length -
+    longerLengthGroupedProjectsByLabel.projectProperties.length
+  );
+};
```
Comment (mecies) [inline src/pages/portfolio/index.tsx]: in `/views/portfolio-page` create `utils.ts` file and move the helper function there

File: src/pages/portfolio/index.tsx [hunk 3]
```diff
@@ -75,27 +76,75 @@ const useStyles = makeStyles((theme) => ({
...
+
+  const otherGroupOfProjects = groupByOtherTab.reduce(
+    (acc, currentGroupProjects) => {
+      return { ...acc, projectProperties: [...currentGroupProjects.projectProperties, ...acc.projectProperties] };
+    },
+    { projectLabel: OtherLabel, projectProperties: [] },
+  );
+  return [...groupByBasicTab, otherGroupOfProjects];
+};
+
```
Comment (mecies) [inline src/pages/portfolio/index.tsx]: also move to `/views/portfolio-page/utils.ts`

File: src/pages/portfolio/index.tsx [hunk 4]
```diff
@@ -75,27 +76,75 @@ const useStyles = makeStyles((theme) => ({
...
-  const filteredProjects = projects.filter((project) => {
-    return projectType !== defaultType ? project.projectLabel === projectType : ' ';
-  });
+  // filter all projects depending on their label
+  const filteredProjects = sortedGroupsOfProjects
+    .filter((project) => {
+      return tabLabelTitle !== defaultNavbarTitle ? project.projectLabel === tabLabelTitle : ' ';
+    })
+    .map((project) => project.projectProperties)
+    .flat(1);
```
Comment (mecies) [inline src/pages/portfolio/index.tsx]: Please notice that this implementation breaks switching between projects while looking at `PortfolioProjectDialog` view,  please consider refactoring this function or refactoring `handleCloseProject`, `handlePreviousProject` and `handleNextProject` not to rely on indices, wdyt? 😄

File: src/pages/portfolio/index.tsx [hunk 5]
```diff
@@ -75,27 +76,75 @@ const useStyles = makeStyles((theme) => ({
...
+  shorterLengthGroupedProjectsByLabel: GroupedProjectsByLabelType,
+): number => {
+  return (
+    shorterLengthGroupedProjectsByLabel.projectProperties.length -
+    longerLengthGroupedProjectsByLabel.projectProperties.length
+  );
+};
+
+const getTabsData = (projects: ProjectType[]) => {
+  const groupedProjects: Record<string, ProjectType[]> = _.groupBy(projects, 'projectLabel');
```
Comment (mecies) [inline src/pages/portfolio/index.tsx]: We could add `lodash` if it was useful application-wide, for a single-use case you don't need a whole library.
Moreover, in my opinion, you don't need a generic function for only handling project labels.

Let's define our own groupByProjectLabel and move it to `/views/portfolio-page/utils.ts`

If you want you can try this, it should work 😄 
I took a generic groupBy method and transformed it to our use-case, lodash bundle size is sooo huge 🍭 
```typescript
function groupByProjectLabel(projects: ProjectType[]): Record<string, ProjectType[]> {
  return projects.reduce((groupedProjects, project) => {
    // eslint-disable-next-line no-param-reassign
    (groupedProjects[project.projectLabel] = groupedProjects[project.projectLabel] || []).push(project);
    return groupedProjects;
  }, {} as { [key: string]: ProjectType[] });
}
```

File: src/pages/portfolio/index.tsx [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import React, { useState } from 'react';
 import { Box, makeStyles } from '@material-ui/core';
 import { graphql } from 'gatsby';
+import * as _ from 'lodash';
```
Comment (mecies) [inline src/pages/portfolio/index.tsx]: This guy has to go 🚀

**Review by mecies — CHANGES_REQUESTED** [review]
Nice job Klara 💪 

The thingy is working nicely, we just need to get rid of one heavy guy 🗿

File: src/pages/portfolio/index.tsx [hunk 7]
```diff
@@ -75,27 +76,75 @@ const useStyles = makeStyles((theme) => ({
   },
 }));
 
+const numberOfMaxTabs = 3;
+const OtherLabel = 'More';
```
Comment (KonradSzwarc) [inline src/pages/portfolio/index.tsx]: Why is it written in CamelCase?

File: src/pages/portfolio/index.tsx [hunk 8]
```diff
@@ -75,27 +76,75 @@ const useStyles = makeStyles((theme) => ({
   },
 }));
 
+const numberOfMaxTabs = 3;
```
Comment (KonradSzwarc) [inline src/pages/portfolio/index.tsx]: Maybe `maxNumberOfTabs` would be better

File: src/pages/portfolio/index.tsx [hunk 9]
```diff
@@ -75,27 +76,75 @@ const useStyles = makeStyles((theme) => ({
...
+    },
+    { projectLabel: OtherLabel, projectProperties: [] },
+  );
+  return [...groupByBasicTab, otherGroupOfProjects];
+};
+
 const PortfolioPage: FC<{ data: ProjectGQL }> = ({ data }) => {
   const classes = useStyles();
-  const defaultType = 'All';
+  const defaultNavbarTitle = 'All';
```
Comment (KonradSzwarc) [inline src/pages/portfolio/index.tsx]: It's constant so you can move it outside of the component

File: src/pages/portfolio/index.tsx [hunk 10]
```diff
@@ -75,27 +76,75 @@ const useStyles = makeStyles((theme) => ({
...
   const developerProfile = useDeveloperProfile();
   const [selectedProject, setSelectedProject] = useState(-1);
   const [navbarTitle, setNavbarTitle] = useState(0);
-  const projectLabels = [defaultType, ...new Set(projects.map((project) => project.projectLabel))];
+  const sortedGroupsOfProjects = getTabsData(projects);
+  const projectsLabels: string[] = [defaultNavbarTitle, ...sortedGroupsOfProjects.map((object) => object.projectLabel)];
 
   const getNavbarTitle = (type: number) => {
-    const title: Record<number, string> = { ...projectLabels };
+    const title: Record<number, string> = { ...projectsLabels };
```
Comment (KonradSzwarc) [inline src/pages/portfolio/index.tsx]: Why do you need to convert the array to an object here?

**Review by KonradSzwarc — CHANGES_REQUESTED** [review]

File: src/pages/portfolio/index.tsx [hunk 11]
```diff
@@ -77,43 +78,61 @@ const useStyles = makeStyles((theme) => ({
...
 
-  // if is first project, choose last project
+  // choose previous project or if is first label, choose last label
   const handlePreviousProject = (index: number) => {
-    setSelectedProject(index === 0 ? projectData.projects.length - 1 : index - 1);
+    return index === 0 ? previousNavbarTitle() : setSelectedProject(index - 1);
   };
 
-  // if is last project, choose first project
+  // choose next project or if is last label, choose first label
```
Comment (KonradSzwarc) [inline src/pages/portfolio/index.tsx]: In my opinion, the name of the function is really meaningful and we don't need a comment here

File: src/pages/portfolio/index.tsx [hunk 12]
```diff
@@ -77,43 +78,61 @@ const useStyles = makeStyles((theme) => ({
...
+    setSelectedProject(0);
   };
 
   // no project will have index equal to -1 therefore no project will be selected
   const handleCloseProject = () => {
     setSelectedProject(-1);
   };
 
-  // if is first project, choose last project
+  // choose previous project or if is first label, choose last label
```
Comment (KonradSzwarc) [inline src/pages/portfolio/index.tsx]: In my opinion, the name of the function is really meaningful and we don't need a comment here

File: src/pages/portfolio/index.tsx [hunk 13]
```diff
@@ -77,43 +78,61 @@ const useStyles = makeStyles((theme) => ({
...
   };
+  const tabLabelTitle = getNavbarTitle(navbarNumberTitle);
 
-  const projectType = getNavbarTitle(navbarTitle);
-
-  const filteredProjects = projects.filter((project) => {
-    return projectType !== defaultType ? project.projectLabel === projectType : ' ';
-  });
+  // filter all projects depending on their label
+  const filteredProjects = sortedGroupsOfProjects
```
Comment (KonradSzwarc) [inline src/pages/portfolio/index.tsx]: Maybe just name the variable `projesctFilteredByLabel` and remove the comment above

File: src/pages/portfolio/index.tsx [hunk 14]
```diff
@@ -77,43 +78,61 @@ const useStyles = makeStyles((theme) => ({
...
   const [selectedProject, setSelectedProject] = useState(-1);
-  const [navbarTitle, setNavbarTitle] = useState(0);
-  const projectLabels = [defaultType, ...new Set(projects.map((project) => project.projectLabel))];
-
+  const [navbarNumberTitle, setNavbarNumberTitle] = useState(0);
+  const sortedGroupsOfProjects = getTabsData(projects);
+  const projectsLabels: string[] = [...sortedGroupsOfProjects.map((object) => object.projectLabel)];
   const getNavbarTitle = (type: number) => {
-    const title: Record<number, string> = { ...projectLabels };
+    const title = projectsLabels;
```
Comment (KonradSzwarc) [inline src/pages/portfolio/index.tsx]: Why do you need to assign `projectsLabels` to a variable here?

File: src/pages/portfolio/utils.ts [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    longerLengthGroupedProjectsByLabel.projectProperties.length
+  );
+};
+
+const groupBy = (projects: ProjectType[]) => {
+  return projects.reduce((groupedProjects, project) => {
+    // eslint-disable-next-line no-param-reassign
+    (groupedProjects[project.projectLabel] = groupedProjects[project.projectLabel] || []).push(project);
+    return groupedProjects;
+  }, {} as { [key: string]: ProjectType[] });
```
Comment (KonradSzwarc) [inline src/pages/portfolio/utils.ts]: Wouldn't it work if you write `reduce<{ [key: string]: ProjectType[] }>(...)` instead of this type casting?

**Review by KonradSzwarc — CHANGES_REQUESTED** [review]

**Review by mecies — CHANGES_REQUESTED** [review]
Agree with @KonradSzwarc  suggestions, please also move `src/pages/portfolio/utils.ts` to => `/src/views/portfolio-page/utils.ts`
Only page components can be stored in the `/pages` directory because Gatsby creates routes out of them automatically, util file in pages directory will break the production build 😄

File: src/views/portfolio-page/utils.ts [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+  longerLengthGroupedProjectsByLabel: GroupedProjectsByLabelType,
+  shorterLengthGroupedProjectsByLabel: GroupedProjectsByLabelType,
+): number => {
+  return (
+    shorterLengthGroupedProjectsByLabel.projectProperties.length -
+    longerLengthGroupedProjectsByLabel.projectProperties.length
+  );
+};
+
+const groupBy = (projects: ProjectType[]) => {
```
Comment (mecies) [inline src/views/portfolio-page/utils.ts]: I suggest to name it `groupByProjectLabel` because it has `projectLabel` hardcoded in the function 😄

**Review by mecies — APPROVED** [review]
Good job 🚀 
Looks good to me ✅

**Review by KonradSzwarc — APPROVED** [review]

**Pull request merged**